kafka-connector/src/com/axonivy/connector/kafka/KafkaStartEventBean.java

```diff
@@ -145,123 +106,121 @@ protected boolean isSynchronous() {
+		private ArrayList<AsyncRequest> asyncRequests = new ArrayList<>();
+		private Properties configuration;
 		private KafkaTopicConsumerSupplier<?, ?> topicConsumerSupplier;
+		private boolean synchronous;
 		private Duration pollTimeout;
+		private KafkaConsumer<?, ?> consumer;
 
-		private KafkaConsumerRunnable(Properties properties, KafkaTopicConsumerSupplier<?, ?> topicConsumerSupplier, Duration pollTimeout) {
-			this.properties = properties;
+		public synchronized void start(Properties configuration, KafkaTopicConsumerSupplier<?, ?> topicConsumerSupplier, boolean synchronous, Duration pollTimeout) {
+			this.configuration = configuration;
 			this.topicConsumerSupplier = topicConsumerSupplier;
+			this.synchronous = synchronous;
 			this.pollTimeout = pollTimeout;
 		}
+		
+		public void stop() {
+			if(consumer != null) {
+				consumer.wakeup();
+			}
+		}
 
-		@Override
-		public void run() {
+		private void run() {
 			try {
-				synchronous = isSynchronous();
-				log().debug("Handle Kafka messages synchronously: {0}", synchronous);
-
-				consumer = topicConsumerSupplier.apply(properties, getTopicPattern());
-				log().debug("Got consumer {0} which subscribed to Kafka topic pattern: {1}", consumer, getTopicPattern());
+				try {
+					log().debug("Handle Kafka messages synchronously: {0}", synchronous);
+					
+					createConsumer();
 
-				while(!consumerExecutor.isShutdown()) {
-					try {
+					while(!Thread.currentThread().isInterrupted()) {
 						ConsumerRecords<?, ?> records = consumer.poll(pollTimeout);
 						for (ConsumerRecord<?, ?> record : records) {
 							log().debug("Received record, topic: {0} offset:{1} key: {2} val: {3} record: {4}",
 									record.topic(), record.offset(), record.key(), record.value(), record);
-							KafkaWorkerRunnable kafkaWorkerRunnable = new KafkaWorkerRunnable(workerExecutor, record, consumer);
-							if(synchronous) {
-								kafkaWorkerRunnable.run();
-							}
-							else {
-								workerExecutor.execute(kafkaWorkerRunnable);
-							}
+							
+							startProcess(record, consumer, synchronous);
 						}
-					} catch (WakeupException e) {
-						log().info("Consumer was woken up. This is ok, when there is a shutdown.");
+						consumeAsyncRequests();
 					}
+				} finally {
+					closeConsumer();
 				}
+			} catch (WakeupException | InterruptException e) {
+				log().info("Consumer was woken up. This is ok, when there is a shutdown.");
+				return;
 			}
 			catch(Exception e) {
 				log().error("Exception while listening for topic, closing consumer.", e);
-			}
-			finally {
-				if(consumer != null) {
-					consumer.close();
-				}
+				throw new RuntimeException("Exception while listening for topic.", e);
 			}
 		}
 
-		public void wakeup() {
-			if(consumer != null) {
-				consumer.wakeup();
+		
+		private void startProcess(ConsumerRecord<?, ?> record, KafkaConsumer<?, ?> consumer2, boolean synchronous) {

```

Comment:
consumer2 is obsolete

Reply:
Good catch. Don't know why the compiler did not warn me about this ...